### ISSUE TITLE
[3.13 CANARY] [REGRESSION] observers fire the first time a computed is accessed

### ISSUE THREAD

**Issue by runspired** [emberjs/ember.js#18225]
**Setup**

```ts
const Thing = EmberObject.extend({
  someProp: computed(function() {
    return { name: 'foo' }
  })
});

// ....
// an instance of Thing

addObserver(thing, 'someProp.name', function() {
    throw new Error(`unexpected observer firing`);
});

// access 'someProp' for the first time
thing.get('someProp')
```

Prior to 3.13 canary (and the tracked work) this observer would not fire on the initial access for `someProp`, now it does.

A reproduction can be observed here: https://github.com/emberjs/data/blob/99f95a4c832d28b2c0e6af8903edfb52ee11f7cf/packages/-ember-data/tests/unit/model/rollback-attributes-test.js#L369

**Comment by rwjblue** [comment]
I don't fully understand why you've categorized this as a regression. I'd absolutely expect the behavior you are describing. Prior to being retrieved the first time, the computed had no value, once retrieved the observer fires because the value that it is observing has changed (went from nothing to something).

**Comment by pzuraq** [comment]
It's more that the value before was _indeterminate_ since it hadn't been retrieved - it was whatever the initial value was. Moreso, this is just not how observers behaved before, as evidenced by their tests. 

I'm trying to think of a way we can potentially match the old behavior, but if it's not possible, we could also debate whether or not this would be an acceptable change in behavior.

**Comment by runspired** [comment]
@rwjblue it's a regression because we now fire for a scenario we did not before, and also because firing in this scenario may have significant perf implications for certain kinds of work, including array-proxy creation

**Comment by runspired** [comment]
FWIW I am open to the idea that this change in behavior is acceptable, I opened the issue because it is a change in behavior that we didn't expect and has ramifications.

**Comment by GavinJoyce** [comment]
FYI, I believe that this change in observer behavior is causing our app to hang with 100% CPU usage (I've seen others report similar behavior in `#topic-octane-migration` too, might be related). I'll work on adding a failing test in an ember.js PR today

**Comment by rwjblue** [comment]
@gavinjoyce - I _think_ you mentioned that updating to 3.13.0-beta.3 addressed that issue for y'all. Can you confirm?

**Comment by GavinJoyce** [comment]
@rwjblue yes, we were seeing infinite invalidation cycles with observers in `3.13.0-beta.2` which have been fixed with `3.13.0-beta.3`. We have a large test suite testing the same addon which is 🍏with `3.13.0-beta.3`

**Comment by bantic** [comment]
We've seen something that might be related when testing out Ember-Table with Ember 3.13. I've filed a lot more detail in an issue at Ember-Table: https://github.com/Addepar/ember-table/issues/775

What we've noticed is that reordering the columns of an Ember-Table table takes an exceptionally long time when using Ember 3.13 (about 100x slower than Ember 3.12). There are a lot of computeds and observers in Ember-Table's `ColumnTree` internal model that it uses to keep handle column changes (such as resizing, reordering, etc.). At first I thought that reordering a column triggered an infinite loop, but when testing on a smaller table I realized it just took a very long time. The majority of the work seems to happen due to a call to `Array#replace` in Ember-Table that ends up calling `arrayContentWillChange` in ember-metal that then calls `flushSyncObservers`. The loop over `activeObservers` seems to get hit a lot. 

In addition, in the call stack I see plenty of nested calls to `getPossibleMandatoryProxyValue` and `CPGETTER_FUNCTION` that, a few items down the stack, end up calling `flushSyncObservers` again. This repeats many times.

I've created a [demo repo](https://github.com/bantic/ember-table-3-13-demo) with a repro, as well as a deployed ember app that also shows the issue: https://et-3-13-bug.surge.sh.

**Comment by pzuraq** [comment]
@bantic unfortunately this is expected. It's not really possible for synchronous observers to be performant in an autotracking based world, and when you have a lot of active observers at once, this will result in _every_ change hitting those observers.

The solution is to switch them to asynchronous observers. This should be possible via the polyfill, but I'm not sure how far back that will work (cc @rwjblue). This may also be a case where we can provide an alternative implementation of the class which does not use observers at all using `ember-compatibility-helpers` in 3.13+

**Comment by mixonic** [comment]
@pzuraq as the performance issue only comes up in 3.13 and [async observers are available in 3.13](https://api.emberjs.com/ember/3.13/classes/Observable/methods/addObserver?anchor=addObserver) it seems ember-table should simply need to detect when in Ember 3.13+ and in that case use async observers. Presuming the implementation in the lib doesn't require sync flushing it sounds like that would easily recoup the performance loss.

**Comment by pzuraq** [comment]
Ah, yes that also seems like it would be a good strategy. It definitely should be much more performant (possibly even more performant than originally, since async observers only flush once per render).

**Comment by bantic** [comment]
@pzuraq @mixonic Thanks for the suggestion. I'm working on a remediation for Ember Table that does as you suggest, by using an internal util that will add the `sync=false` argument for Ember 3.13+.

So far it looks like that does solve the issue. However this uncovered a new problem, which is that, when using plain-old javascript arrays for column definitions, the column property change notification (after `splice`-ing the column on reorder) isn't propagating the way it does when either a) `columns` is an Ember Array or b) on Ember 3.12 and below. I filed a related issue https://github.com/Addepar/ember-table/issues/776. Is that also an expected change when using async observers? Is there another way to propagate a property-change notification for an array's content so that async observers will kick in?

**Comment by sandstrom** [comment]
I'm doing some issue gardening 🌱🌿 🌷 and came upon this issue. Since it's quite old I just wanted to ask if this is still relevant? If it isn't, maybe we can close this issue? (also because observers are gradually being phased out of Ember)

By closing some old issues we reduce the list of open issues to a more manageable set.

**Comment by shama** [comment]
This issue is still relevant to us. We are trying to upgrade our app from 3.12 to 3.13. We have a Shared Library that when consumed by Ember, it heavily relies on observers. With `"default-async-observers": false`, our tests will run so slow they crash the browser. With `"default-async-observers": true`, the tests run but it introduces a lot of timing issues.

We would love to migrate away from observers but tracked properties are not added until Ember 3.13.

Does anyone have any suggestions for migrating to 3.13? Or is there a way to make observers work like they did in Ember 3.12?

**Comment by kategengler** [comment]
I'm going to close this issue. Opting into async observers has been an option since 3.13, and the default since 4.0. 

We've long planned to deprecate and remove observers and ideally this will happen in the 7.x timeframe https://github.com/emberjs/rfcs/pull/1115
